File: 백준/Gold/2110. 공유기 설치/공유기 설치.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


int n, c;
int x;
int lt, rt;
int mx = 0;//최대
int mn = 1000000000;//최소
int ans = 1;
vector<int> v;

int main() {
	cin >> n;
	cin >> c;
	while (n--) {
		cin >> x;
		if (x > mx) {
			mx = x;
		}
		if (mn > x) mn = x;
		v.push_back(x);
	}

	sort(v.begin(), v.end());

	lt = 1;
	rt = (mx - mn) / (c - 1); //처음부터 마지막까지 거리를 c-1로 나눴을때 최대

	

	while(rt >= lt) {

	int cnt = 0;

	//cout << lt << ' ' << rt << endl;
	int mid = (rt +lt) / 2;
	//cout << "mid : " << mid << endl;
	int temp = mn;
	
	for (int i = 0; i < v.size(); ++i) {
		if (v[i] - temp >= mid) {
			//cout << v[i] << " "<< temp << endl;
			temp = v[i];
			cnt++;
		}
	}
	
	//cout << "cnt : " << cnt << "  c-1:  " << c - 1 << endl;

	if (cnt >= c-1) { 
		ans = mid;
		lt = mid + 1; 
	
	}
	else {
		rt = mid - 1;
	}

	
	}

	cout << rt;


}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
n, c = map(int, input().split())

loc = []
for i in range(n):
    loc.append(int(input()))

loc.sort()

start =1
end = loc[-1]-loc[0]
ans =1 
def binary_search(arr, start, end):
    
    while start <= end:
        temp = arr[0]
        cnt = 0
        mid = (start + end)//2

        for i in range(1, len(arr)):   
            if(arr[i]- temp >= mid):
                temp = arr[i]
                cnt= cnt +1

        if(cnt < c-1):
            end = mid-1

        else:
            ans = mid
            start = mid +1

    print(ans)

binary_search(loc, start, end)    
